core/src/main/scala/kafka/utils/ZkUtils.scala

```diff
@@ -256,19 +251,41 @@ class ZkUtils(val zkClient: ZkClient,
   }
 
   /**
-   * Register brokers with v2 json format (which includes multiple endpoints).
+   * Register brokers with v3 json format (which includes multiple endpoints and rack).
```

Comment:
@junrao Thanks for explaining. If we want to do it this way, I think we need to add a comment explaining it because it's not obvious and it's inconsistent with how we did the V1 to V2 change. Having said that, I am not sure if it's worth doing this now for the following reasons:

* It will probably be a long time before we can actually drop support for the old consumer and admin tools in the broker (I think 3 non-bug fix releases is the absolute minimum, but it will probably be longer)
* It's inconsistent with how the parser code works (for compatibility as you mentioned). Without changing the parser code, older brokers will break if we suddenly remove fields.
* It doesn't seem to buy us anything in terms of what gets stored (people will bump inter.broker.protocol as part of the upgrade and then V3 will be stored anyway) for the time when we need

It seems to me that when we have a concrete plan to clean this up, we can add the necessary code for both writing and parsing. It will probably take a few releases before we can actually delete the fields, but that's probably OK.